include/sparrow/array_data_factory.hpp

```diff
@@ -0,0 +1,359 @@
+            .child_data = {},
+            .dictionary = nullptr
+        };
+    }
+
+    template <typename V>
+    struct ValuesAndIndexes
+    {
+        template <mpl::constant_range R>
+            requires std::same_as<std::ranges::range_value_t<R>, std::remove_const_t<V>>
+        explicit ValuesAndIndexes(R&& range)
+        {
+            ranges_to_vec_and_indexes<R>(range, *this);
+        }
+
+        std::vector<std::reference_wrapper<V>> values;
+        std::vector<size_t> indexes;
+    };
+
+    template <mpl::constant_range R>
+    void
+    ranges_to_vec_and_indexes(R&& range, ValuesAndIndexes<const std::ranges::range_value_t<R>>& values_and_indexes)
+    {
+        using T = const std::ranges::range_value_t<R>;
+        std::unordered_map<std::reference_wrapper<T>, size_t, reference_wrapper_hasher, reference_wrapper_equal>
+            set_index;  // TODO: To replace with a flat_map/another implementation when available, for
+                        // performance reasons
+        for (const auto& value : range)
+        {
+            set_index.emplace(std::cref(value), 0);
+        }
+
+        for (size_t i = 0; auto& [_, value] : set_index)
+        {
+            value = i;
+            ++i;
+        }
+
+        values_and_indexes.indexes.reserve(std::ranges::size(range));
+
+        std::ranges::transform(
+            range,
+            std::back_inserter(values_and_indexes.indexes),
+            [&set_index](const T& value)
+            {
+                return set_index[std::cref(value)];
+            }
+        );
+
+        values_and_indexes.values.reserve(set_index.size());
+
+        std::ranges::transform(
+            set_index,
+            std::back_inserter(values_and_indexes.values),
+            [](const auto& pair)
+            {
+                return pair.first;
+            }
+        );
+    }
+
+    template <typename T>
+    array_data make_array_data_for_dictionary_encoded_layout()
+    {
+        return {
+            .type = data_descriptor(arrow_type_id<std::uint64_t>()),
+            .length = 0,
+            .offset = 0,
+            .bitmap = {},
+            .buffers = {array_data::buffer_type(sizeof(std::int64_t), 0)},
+            .child_data = {},
+            .dictionary = value_ptr<array_data>(make_array_data_for_variable_size_binary_layout<T>())
+        };
+    }
+
+    template <mpl::constant_range ValueRange>
+    array_data make_array_data_for_dictionary_encoded_layout(
+        ValueRange&& values,
+        const array_data::bitmap_type& bitmap,
+        std::int64_t offset
+    )
+    {
+        SPARROW_ASSERT_TRUE(values.size() == bitmap.size());
+        SPARROW_ASSERT_TRUE(std::cmp_greater_equal(values.size(), offset));
+
+        const ValuesAndIndexes<const std::ranges::range_value_t<ValueRange>> vec_and_indexes{values};
+        const auto& indexes = vec_and_indexes.indexes;
+        static const auto create_buffer = [&indexes]()
```

Comment:
Why making this lambda static? I'm afraid it initialized the capture only once, leading to crashes at runtime during a second call to `make_array_data_for_dictionary_encoded_layout`